IsRebuildAll and Form showing monthly totals

ceddlyburge opened the pull request:
HI Daniel

This pull request introduces a new variable "IsRebuildAll", which is output to the buildmonitor window and saved with the JSON.

It also includes the changes from the previous pull request, which were to show a form with a breakdown of total compile time by month and by solution.

Cheers

Cedd

Commit:
Adding a menu item that displays a form and check that it works.

Commit:
Add a simple test to get the total build time for a single solution build

Commit:
Add tests for calculating build time per solution and per solution month

Commit:
Write build times out to the ressoftware database

Commit:
Updating version number

Commit:
Add tests / functionality to provide a list of available solutionmonths

Commit:
Add test and code for availablesolutions

Commit:
Start adding form to show breakdown of buildtime for solution by month, add some tests based on the problems encountered

Commit:
move ui code in to window and calculation ish code in to calculator

Commit:
Making the UI look a bit more acceptable

Commit:
Formatting months as MMM

Commit:
£Fix conflict by preferring version 1.5

Commit:
Removing RES specific code

Commit:
Adding an IsRebuildAll flag to output

Commit:
Setting isrebuildall flag at end of build so that the buildbegin event will have been acted on

Commit:
Fixing up test compilation error

Commit:
Fixing up tests

Commit:
Merge remote-tracking branch 'origin/VS2015'

Commit:
changing name of button on ui and setting a min size for forn

Commit:
Removing res specific code

Commit:
Remove backup directory

vinntreus commented on BuildMonitor/BuildMonitor.csproj:
```diff
@@ -62,15 +63,22 @@
   </ItemGroup>
   <ItemGroup>
+    <None Include="cedd.snk" />
     <None Include="key.pfx" />
```
My memory of signing is a bit fuzzy, but is the key.pfx still needed?

ceddlyburge replied:
It depends. The snk is used for strong-naming, which uses a key pair to uniquely identify an assembly. The pfx is for code signing, which is a similar process but one that is intended to prevent malicious tampering with assemblies that are distributed publicly. Using both on an assembly both strong-names AND signs it.

It would probably be best to compile and release it using your original pfx, but I can't do this as I don't have the password.

vinntreus commented on BuildMonitor/Domain/ISolutionBuild.cs:
```diff
@@ -6,6 +6,8 @@ public interface ISolutionBuild : ITimer, IPersistable
     {
         DateTime Started { get; }
         ISolution Solution { get;  }
+        bool IsRebuildAll { get; set; } // bit of a shame to make this mutable
```
If it's a shame then why make it mutable? ;-) I don't have a big problem with making stuff mutable, but am not a big fan of comments that don't really add any value.

ceddlyburge replied:
This is because we don't know it when we construct the object, but I'm happy to remove the comment.

vinntreus commented on BuildMonitor/LocalData/BuildTimes.cs:
```diff
@@ -0,0 +1,84 @@
+        // this is for the UI, in MVC land it would be a ViewModel, but that seems a bit excessive here.
+        public IEnumerable<ExpandoObject> SolutionMonthTable()
+        {
+            var table = new List<ExpandoObject>();
```
Why using an ExpandoObject when it looks like it's a Dictionary<string, object>?

ceddlyburge replied:
This is required in order to use it as the ItemsSource property on a DataGrid.

vinntreus commented on BuildMonitor/LocalData/SolutionMonth.cs:
```diff
@@ -0,0 +1,49 @@
+﻿using System;
```
-- Just a marker to know where I shall continue to review --

Commit:
Remove comment about mutability

Commit:
Upgrade to work with Visual Studio 2017

Commit:
Upgrade to work with Visual Studio 2017

ceddlyburge committed:
Update README.markdown

Commit:
log to kl-sql-005

edespong committed:
Removed references causing compile error.

Also removed the VS upgrade log, which seems unnecessary.

ceddlyburge committed:
Remove envdte projects, fixes #1

The solution now builds in vs 2017

ceddlyburge committed:
Comment out office specific database saving code

edespong committed:
Added options page with output file path

Settings is no longer a static class in order to be able to create it with a WritableSettingsStore to persist data. Instead, it has a static instance.

edespong committed:
Write to output when changing settings

Broke out the output window code to separate class.

Also removed the IntExtension which can be replaced by a standard string constructor.

edespong committed:
Merged with master

edespong committed:
Merge branch 'master' into Settings

ceddlyburge committed:
Allow location of json file to be set in an environment variable #2

Commit:
Ignore visual studio nuser settings

Commit:
Add appveyor configuration

Commit:
Fix build target

Commit:
Remove pack msbuild argument

Commit:
Use nunit 2 console runner

Commit:
Fix test dll path

Commit:
Update readme

Commit:
Attempt to add coverage

Commit:
Fix nunit command line

Commit:
Add coverage.xml as artifact to debug

Commit:
Build in debuf configuration to allow coverage to work

ceddlyburge committed:
Merge branch 'master' into Settings

ceddlyburge committed:
Merge branch 'master' into Settings

ceddlyburge committed:
Merge pull request #3 from edespong/Settings

Add ability to change output path

Commit:
Support Visual Studio 2019

It switches to Async initialization to improve load time performance.
Probably more things could be done before switching to the main ui
thread, but its still an improvment.

I haven't got the tests to run when opening the solution in vs 2019
yet

Commit:
Upgrade to nunit 3

The nunit 2 tests wouldn't run in vs 2019, despite doing all the
things you are supposed to do.

Upgrading to 3 is probably a good idea anyway though

Commit:
Use nunit3 during build

Commit:
Update nunit 3 test runner in build script

Commit:
Fix problem saving to database

The problem happend because the solution name was null. I have made
sure that this can't happen, and an empty string is now used instead.

I have also found the reason why the solution name was null and
fixed that. The Opened event doesn't fire when a solution is
opened when starting vs 2019, as I think we only connect to the
events after the solution is already open. To get round this I call
the event manually during initialize.

ceddlyburge committed:
Add visual studio support to readme
